cartographer/mapping/internal/range_data_collator.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/internal/range_data_collator.h"
+
+#include <memory>
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/mapping/local_slam_result_data.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace mapping {
+
+sensor::TimedPointCloudOriginData RangeDataCollator::AddRangeData(
+    const std::string& sensor_id,
+    const sensor::TimedPointCloudData& timed_point_cloud_data) {
+  CHECK(expected_sensor_ids_.count(sensor_id) != 0);
+  // TODO(gaschler): These two cases can probably be one.
+  if (id_to_pending_data_.count(sensor_id) != 0) {
+    current_start_ = current_end_;
+    // When we have two messages of the same sensor, move forward the older of
+    // the two (do not send out current).
+    current_end_ = id_to_pending_data_.at(sensor_id).time;
+    auto result = CropAndMerge();
+    id_to_pending_data_.emplace(sensor_id, timed_point_cloud_data);
+    return result;
+  }
+  id_to_pending_data_.emplace(sensor_id, timed_point_cloud_data);
+  if (expected_sensor_ids_.size() != id_to_pending_data_.size()) {
+    return {};
+  }
+  current_start_ = current_end_;
+  // We have messages from all sensors, move forward to oldest.
+  common::Time oldest_timestamp = common::Time::max();
+  for (const auto& pair : id_to_pending_data_) {
+    oldest_timestamp = std::min(oldest_timestamp, pair.second.time);
+  }
+  current_end_ = oldest_timestamp;
+  return CropAndMerge();
+}
+
+sensor::TimedPointCloudOriginData RangeDataCollator::CropAndMerge() {
```

Comment:
```
namespace {
sensor::TimedPointCloudOriginData CropAndMerge(const common::Time& start, const common::Time& end,  std::map<std::string, sensor::TimedPointCloudData>* id_to_pending_data) {

}
}  // namespace

Reply:
I'd prefer to keep this a method because it needs all variables of RangeDataCollator expect  expected_sensor_ids.